File: backend/src/main/kotlin/com/predata/backend/domain/Order.kt
package com.predata.backend.domain

import jakarta.persistence.*
import java.math.BigDecimal
import java.time.LocalDateTime

@Entity
@Table(name = "orders")
data class Order(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "order_id")
    val id: Long? = null,

    @Column(name = "member_id", nullable = false)
    val memberId: Long,

    @Column(name = "question_id", nullable = false)
    val questionId: Long,

    @Enumerated(EnumType.STRING)
    @Column(name = "order_type", nullable = false)
    val orderType: OrderType,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val side: OrderSide,

    @Column(precision = 4, scale = 2, nullable = false)
    val price: BigDecimal,

    @Column(nullable = false)
    val amount: Long,

    @Column(name = "remaining_amount", nullable = false)
    var remainingAmount: Long,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: OrderStatus = OrderStatus.OPEN,

    @Column(name = "created_at", nullable = false, updatable = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),

    @Column(name = "updated_at", nullable = false)
    var updatedAt: LocalDateTime = LocalDateTime.now()
)

enum class OrderType {
    LIMIT,   // 지정가 주문 (오더북 적재)
    MARKET   // 시장가 IOC 주문 (즉시 체결, 미체결분 취소)
}

enum class OrderSide {
    YES, NO
}

enum class OrderStatus {
    OPEN, FILLED, PARTIAL, CANCELLED
}

File: src/lib/api/order.ts
import { apiRequest } from './core';

export interface OrderBookLevel {
  price: number;
  amount: number;
  count: number;
}

export interface OrderBookData {
  questionId: number;
  bids: OrderBookLevel[];
  asks: OrderBookLevel[];
  lastPrice?: number;
  spread?: number;
}

export interface CreateOrderRequest {
  questionId: number;
  side: 'YES' | 'NO';
  // MARKET orders don't send price (server fills at best price), LIMIT orders require price
  price?: number;
  amount: number;
  direction?: 'BUY' | 'SELL'; // BUY: Buy (USDC deposit), SELL: Sell (position collateral), defaults to BUY
  orderType?: 'LIMIT' | 'MARKET'; // Order type, defaults to LIMIT
}

export interface CreateOrderResponse {
  success: boolean;
  message?: string;
  orderId?: number;
  filledAmount: number;
  remainingAmount: number;
}

export interface OrderData {
  orderId: number;
  memberId: number;
  questionId: number;
  side: 'YES' | 'NO';
  direction: 'BUY' | 'SELL';
  price: number;
  amount: number;
  remainingAmount: number;
  status: 'OPEN' | 'FILLED' | 'PARTIAL' | 'CANCELLED';
  createdAt: string;
  updatedAt: string;
}

export const orderApi = {
  getOrderBook: (questionId: number) =>
    apiRequest<OrderBookData>(`/api/questions/${questionId}/orderbook`),

  createOrder: (data: CreateOrderRequest) =>
    apiRequest<CreateOrderResponse>('/api/orders', {
      method: 'POST',
      body: JSON.stringify(data),
    }),

  cancelOrder: (orderId: number) =>
    apiRequest<{ success: boolean; message?: string; refundedAmount?: number }>(
      `/api/orders/${orderId}`,
      { method: 'DELETE' }
    ),

  getActiveOrders: () =>
    apiRequest<OrderData[]>('/api/orders/me'),

  getOrdersByQuestion: (questionId: number) =>
    apiRequest<OrderData[]>(`/api/orders/me/question/${questionId}`),
};
